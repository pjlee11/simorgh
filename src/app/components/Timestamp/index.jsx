import React from 'react';
import styled from 'styled-components';
import { number } from 'prop-types';
import { T_BREVIER } from '../../lib/constants/typography';
import {
  C_RHINO,
  GEL_SPACING_DBL,
  FF_NEWS_SANS_REG,
} from '../../lib/constants/styles';

// if the date is invalid return null - https://stackoverflow.com/questions/1353684/detecting-an-invalid-date-date-instance-in-javascript#answer-1353711
const isValidDateTime = dateTime => !isNaN(dateTime); // eslint-disable-line no-restricted-globals

const formatDateTime = dateObj => {
  const fullYear = dateObj.getFullYear();
  const monthTwoDigit = dateObj.toLocaleDateString('en-GB', {
    month: '2-digit',
  });
  const dayTwoDigit = dateObj.toLocaleDateString('en-GB', {
    day: '2-digit',
  });

  return `${fullYear}-${monthTwoDigit}-${dayTwoDigit}`;
};

const formatTimestamp = dateObj => {
  const fullYear = dateObj.getFullYear();
  const monthLong = dateObj.toLocaleDateString('en-GB', {
    month: 'long',
  });
  const dayNumeric = dateObj.toLocaleDateString('en-GB', {
    day: 'numeric',
  });

  return `${dayNumeric} ${monthLong} ${fullYear}`;
};

const StyledTimestamp = styled.div`
  ${T_BREVIER};
  color: ${C_RHINO};
  font-family: ${FF_NEWS_SANS_REG};
  padding-bottom: ${GEL_SPACING_DBL};
`;

const Timestamp = ({ timestamp }) => {
  const dateObj = new Date(timestamp);

  if (!isValidDateTime(dateObj)) {
    return null;
  }

  return (
    <StyledTimestamp>
      <time dateTime={formatDateTime(dateObj)}>{formatTimestamp(dateObj)}</time>
    </StyledTimestamp>
  );
};

Timestamp.propTypes = {
  timestamp: number.isRequired,
};

export default Timestamp;
