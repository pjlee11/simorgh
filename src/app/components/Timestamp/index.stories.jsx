import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import Timestamp from './index';

storiesOf('Timestamp', module).add('default', () => (
  <Timestamp timestamp={1530947227000} />
));
