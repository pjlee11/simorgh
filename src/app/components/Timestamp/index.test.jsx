import React from 'react';
import { isNull, shouldMatchSnapshot } from '../../helpers/tests/testHelpers';
import Timestamp from './index';

const timestamp = 1539969006000; // 19 October 2018
const noLeadingZeroTimestamp = 1530947227000; // 07 July 2018
const invalidTimestamp = 8640000000000001; // A day holds 86,400,000 milliseconds - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date#Description

describe('Timestamp', () => {
  describe('with no data', () => {
    isNull('should return null', <Timestamp />);
  });
  shouldMatchSnapshot(
    'should render without a leading zero on the day',
    <Timestamp timestamp={noLeadingZeroTimestamp} />,
  );
  shouldMatchSnapshot(
    'should render correctly',
    <Timestamp timestamp={timestamp} />,
  );
  shouldMatchSnapshot(
    'should handle an invalid timestamp',
    <Timestamp timestamp={invalidTimestamp} />,
  );
});
