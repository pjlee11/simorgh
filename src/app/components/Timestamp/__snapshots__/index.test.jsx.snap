// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timestamp should handle an invalid timestamp 1`] = `null`;

exports[`Timestamp should render correctly 1`] = `
.c0 {
  font-size: 0.875em;
  line-height: 1rem;
  color: #5A5A5A;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding-bottom: 1rem;
}

@media (min-width:20em) and (max-width:37.4375em) {
  .c0 {
    line-height: 1.125rem;
  }
}

@media (min-width:37.5em) {
  .c0 {
    font-size: 0.8125em;
  }
}

<div
  className="c0"
>
  <time
    dateTime="2018-10-19"
  >
    19 October 2018
  </time>
</div>
`;

exports[`Timestamp should render without a leading zero on the day 1`] = `
.c0 {
  font-size: 0.875em;
  line-height: 1rem;
  color: #5A5A5A;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding-bottom: 1rem;
}

@media (min-width:20em) and (max-width:37.4375em) {
  .c0 {
    line-height: 1.125rem;
  }
}

@media (min-width:37.5em) {
  .c0 {
    font-size: 0.8125em;
  }
}

<div
  className="c0"
>
  <time
    dateTime="2018-07-07"
  >
    7 July 2018
  </time>
</div>
`;
